### PR TITLE
[OpAMP] Harden WebSocket transport

### PR DESCRIPTION
**Pull request by stevejgordon**
Async receive loop, max-message enforcement, ClientWebSocketFactory, and disposal fixes

## Changes

- Replace background Thread with async Task in WsReceiver so the receive loop is cancellable and disposal is deterministic (blocks until the task drains, prevents use-after-free on the socket)
- Remove the insecure DangerousAcceptAnyServerCertificateValidator default on .NET; users who need custom TLS can now set ClientWebSocketFactory
- Add OpAmpClientSettings.ClientWebSocketFactory (new public API) so callers control socket options (headers, TLS, etc.)
- Fix OpAmpClient.Dispose() to actually stop background services and dispose the transport (previously only dispatcher.Dispose() was called)
- Catch WebSocketException/ObjectDisposedException in the receive path so aborted-socket errors propagate cleanly rather than as unobserved faults
- Catch frame-processing exceptions and emit a FrameProcessingFailure event so a single bad frame cannot kill the long-lived connection
- Fix array-pool return bug: was returning the non-pooled receiveBuffer - Fix multi-buffer sequence construction (workingCount off-by-one) - Add OversizedWebSocketMessage and FrameProcessingFailure EventSource events - Add ThrowIfDisposed guards to all public OpAmpClient and WsTransport methods - Add new tests covering boundary conditions, disposal, factory config, invalid/text
  frames, fragmented oversized messages, and error resilience
- Switch StopAsync from CloseAsync to CloseOutputAsync so the graceful close handshake does not conflict with the concurrent ReceiveAsync in WsReceiver; also fixes a potential hang on .NET Framework where CloseAsync blocks waiting for the server close frame
- Fix single-buffer sequence construction: slice receiveBuffer to totalCount so the frame does not include uninitialised tail bytes

## Behaviour Notes

1. **Security fix (breaking for some users): TLS certificate validation is no longer bypassed**

   The .NET-only `DangerousAcceptAnyServerCertificateValidator` default has been removed. Any user
   connecting to an OpAMP server with a self-signed or non-standard TLS cert will need to configure
   `ClientWebSocketFactory` and set `ClientWebSocket.Options` appropriately. HTTP transport is
   unaffected.

2. **New public API: `OpAmpClientSettings.ClientWebSocketFactory`**

   Callers can now supply a factory that returns a pre-configured `ClientWebSocket`. This is the
   migration path for anyone who was relying on the removed TLS bypass, and it enables setting custom
   request headers, keep-alive settings, etc.

3. **`OpAmpClient.Dispose()` now releases transport and service resources**

   Previously `Dispose()` only called `dispatcher.Dispose()` and leaked the `IOpAmpTransport` and
   background services. `Dispose()` now stops all services and disposes the transport. Callers who
   were relying on `Dispose()` to be a no-op (e.g. GC-only cleanup) will now see that resources are
   actively released.

4. **`Dispose()` is not a graceful shutdown — `StopAsync()` should be called first**

   `Dispose()` aborts the `WebSocket` immediately without sending the agent disconnect message. The
   README and XML docs have been updated to make this explicit. Callers who previously called only
   `Dispose()` and expected a clean server-side unregister should switch to `await StopAsync()`
   followed by `Dispose()` (or `using`).

5. **`StopAsync` switches from `CloseAsync` to `CloseOutputAsync`**

   `CloseAsync` internally calls `ReceiveAsync`, which conflicts with the concurrent `ReceiveAsync`
   running in `WsReceiver` and can cause a deadlock or hang on .NET Framework. `CloseOutputAsync`
   only sends the close frame; the receive loop consumes the server's close response naturally and
   exits.

6. **Frame processing errors no longer crash the receive loop**

   A bad or unrecognisable server frame now logs a `FrameProcessingFailure` warning event and moves
   on to the next frame, rather than crashing the `async void` receive thread silently. This changes
   the failure mode from "connection dies silently" to "frame is dropped, warning is logged,
   connection stays up."

7. **Disposal of `WsReceiver` is now synchronous and deterministic**

   `Dispose()` cancels the receive task and blocks (via `GetAwaiter().GetResult()`) until it
   finishes, preventing the socket from being used concurrently after disposal. The previous
   implementation called `Thread.Join()` but never cancelled the cancellation token inside
   `Dispose()`, so if the caller had not cancelled externally the thread would block indefinitely on
   a live socket. The new path always has a cancellation route: `disposeTokenSource` is cancelled
   before the wait, and `WsTransport.Dispose()` calls `webSocket.Abort()` first, which unblocks
   `ReceiveAsync` almost immediately.

8. **Updated `TestHttpServer` and `TestWebSocketServer`**

   The original `RunningServer` used `new Task(async () => { ... })`, which is a known anti-pattern
   where the outer `Task` completes at the first `await` rather than when the async work finishes —
   meaning `Dispose()` was not actually waiting for the listener loop to complete. The fix replaces
   this with `Task.Run(() => this.ListenAsync(...))`, which correctly tracks the full async
   lifecycle. Initialization signalling is also improved: `AutoResetEvent` is replaced with
   `TaskCompletionSource<bool>` and the signal is set before the accept loop begins, so `Start()`
   returns only when the server is unambiguously ready. Shutdown exception handling is consolidated
   into a shared `IsListenerShutdownException` helper. `TestWebSocketServer.RunServer` also gains a
   new overload that exposes the `HttpListenerContext` alongside the `WebSocket`, enabling tests to
   inspect HTTP upgrade request headers (e.g. for verifying custom headers set via
   `ClientWebSocketFactory`); the existing single-argument overload delegates to the new one and is
   unaffected.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by stevejgordon** [comment]
The Sampler test for AWS may be related to changes in the test server, but it runs fine locally.

File: src/OpenTelemetry.OpAmp.Client/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 ## Unreleased
 
+* Harden WebSocket transport: enforce maximum server message size, use a
+  cancellable receive loop with deterministic disposal, add
+  `OpAmpClientSettings.ClientWebSocketFactory`, and remove the insecure default
+  of accepting any server TLS certificate when connecting on .NET.
+  ([#4133](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/4133))
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/CHANGELOG.md]: I'd break the distinct changes into different bullet points.

```
* Harden WebSocket transport
  * Change 1
  * Change 2...
```

File: src/OpenTelemetry.OpAmp.Client/Internal/OpAmpClientEventSource.cs [hunk 2]
```diff
@@ -17,6 +17,8 @@ internal sealed class OpAmpClientEventSource : EventSource
     private const int EventIdOversizedResponseContentLength = 3;
     private const int EventIdOversizedResponseBody = 4;
     private const int EventIdHttpResponseReceived = 5;
+    private const int EventIdOversizedWebSocketMessage = 6;
+    private const int EventIdFrameProcessingFailure = 7;
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Internal/OpAmpClientEventSource.cs]: Once #4130 is merged the event IDs will change.

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsReceiver.cs [hunk 3]
```diff
@@ -8,18 +8,29 @@
...
 namespace OpenTelemetry.OpAmp.Client.Internal.Transport.WebSocket;
 
+/// <summary>
+/// Reads OpAMP WebSocket messages from the server and dispatches them to the frame processor.
+/// </summary>
+/// <remarks>
+/// <para>
+/// Oversized messages: see <see cref="TransportConstants.MaxMessageSize"/> for the size limit
+/// and the note on peak buffered bytes when enforcement happens after each receive operation.
+/// </para>
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsReceiver.cs]: Do we need this in the `///` comments?

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs [hunk 4]
```diff
@@ -1,87 +1,136 @@
...
+    /// Connects the WebSocket and starts receiving server frames.
+    /// </summary>
+    /// <param name="token">A cancellation token for the connect operation.</param>
+    /// <returns>A task that completes when the connection is established and the receive loop has started.</returns>
+    /// <remarks>
+    /// This method is not idempotent: a second call after a successful start throws
+    /// <see cref="InvalidOperationException"/>. If the connection or receive startup fails, internal state is reset
+    /// so that <see cref="Dispose"/> can still be called cleanly. A new <see cref="WsTransport"/> instance
+    /// (and therefore a new <see cref="ClientWebSocket"/>) is required to reconnect.
+    /// </remarks>
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs:70]: Should add the `<exception>` tag to document this too.

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs [hunk 5]
```diff
@@ -1,87 +1,136 @@
...
 
+    /// <summary>
+    /// Connects the WebSocket and starts receiving server frames.
+    /// </summary>
+    /// <param name="token">A cancellation token for the connect operation.</param>
+    /// <returns>A task that completes when the connection is established and the receive loop has started.</returns>
+    /// <remarks>
+    /// This method is not idempotent: a second call after a successful start throws
+    /// <see cref="InvalidOperationException"/>. If the connection or receive startup fails, internal state is reset
+    /// so that <see cref="Dispose"/> can still be called cleanly. A new <see cref="WsTransport"/> instance
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs]: The middle sentence doesn't feel like something callers should need to care about.

File: src/OpenTelemetry.OpAmp.Client/OpAmpClient.cs [hunk 6]
```diff
@@ -144,33 +160,85 @@ public Task SendCustomCapabilitiesAsync(IEnumerable<string> capabilities, Cancel
...
+            if (service is IDisposable disposableService)
+            {
+                disposableService.Dispose();
+            }
+        }
+
+        if (this.transport is IDisposable disposableTransport)
+        {
+            disposableTransport.Dispose();
+        }
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/OpAmpClient.cs:214]: Do we maybe need to consider `IAsyncDisposable` too? Only if any thing we're already using implements one of those.

Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/OpAmpClient.cs:214]: Nothing implements IAsyncDisposable (currently)

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs (hunk 5 above)
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs]: Good shout, restructured this section.

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs (hunk 4 above)
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs:70]: Done + added XML docs to other methods for consistency

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsReceiver.cs (hunk 3 above)
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsReceiver.cs]: No, removed

File: src/OpenTelemetry.OpAmp.Client/CHANGELOG.md (hunk 1 above)
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/CHANGELOG.md]: Done

**Comment by martincostello** [comment]
I _think_ #4121 resolved the Hangfire flakes.

**Comment by stevejgordon** [comment]
> I _think_ #4121 resolved the Hangfire flakes.

Once your PR is merged, I can rebase on top of everything and hopefully we'll see this one go green again!

**Review by martincostello — APPROVED** [review]

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs [hunk 7]
```diff
@@ -1,87 +1,171 @@
...
+            }
+            catch
+            {
+                // Best effort: Abort may throw if the socket is already in a terminal state.
+            }
 
-        this.receiver.Start(token);
+            Interlocked.Exchange(ref this.startState, 0);
+            throw;
+        }
```
[the remark this answers is not in the document]
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs:110]: This wasn't the best fix. Added a richer handling of states.

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs [hunk 8]
```diff
@@ -1,87 +1,171 @@
...
+        if (this.disposed)
+        {
+            return;
+        }
+
+        this.disposed = true;
+
+        this.webSocket.Abort();
         this.receiver.Dispose();
+        this.webSocket.Dispose();
```
[the remark this answers is not in the document]
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs]: Fixed

File: src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs [hunk 9]
```diff
@@ -1,87 +1,171 @@
...
+    /// </summary>
+    /// <param name="token">A cancellation token for the connect operation.</param>
+    /// <returns>A task that completes when the connection is established and the receive loop has started.</returns>
+    /// <remarks>
+    /// This method is not idempotent: a second call after a successful start throws
+    /// <see cref="InvalidOperationException"/>. If connection establishment or receive-loop startup fails,
+    /// the transport aborts the socket and the instance cannot be reused. A new <see cref="WsTransport"/>
+    /// instance (and therefore a new <see cref="ClientWebSocket"/>) is required to reconnect.
+    /// </remarks>
+    /// <exception cref=" ObjectDisposedException">
```
[the remark this answers is not in the document]
Comment (stevejgordon) [reply src/OpenTelemetry.OpAmp.Client/Internal/Transport/WebSocket/WsTransport.cs]: Fixed

**Comment by stevejgordon** [comment]
@martincostello Hopefully ready now!
